File: parent/app/src/main/java/com/guardianstar/parent/network/ApiClient.java
package com.guardianstar.parent.network;

import android.util.Log;

import com.guardianstar.parent.BuildConfig;

import java.io.IOException;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.util.concurrent.TimeUnit;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class ApiClient {

    private static final String TAG = "ApiClient";
    private static final String BASE_URL = BuildConfig.BASE_URL;
    private static Retrofit retrofit;

    /** 快速检测服务器是否可达（TCP 连接测试，5 秒超时） */
    public static boolean isServerReachable() {
        String host = "139.9.176.191";
        int port = 3000;
        try (Socket socket = new Socket()) {
            socket.connect(new InetSocketAddress(host, port), 5000);
            return true;
        } catch (IOException e) {
            Log.w(TAG, "服务器不可达: " + host + ":" + port + " - " + e.getMessage());
            return false;
        }
    }

    /** 日志拦截器：记录完整请求/响应时间线 */
    private static class TimingInterceptor implements Interceptor {
        @Override
        public Response intercept(Chain chain) throws IOException {
            Request request = chain.request();
            long startNs = System.nanoTime();

            Log.i(TAG, "→ " + request.method() + " " + request.url());
            if (request.body() != null && request.body().contentLength() > 0) {
                Log.d(TAG, "  请求体大小: " + request.body().contentLength() + " bytes");
            }

            Response response;
            try {
                response = chain.proceed(request);
            } catch (IOException e) {
                long ms = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNs);
                Log.e(TAG, "✗ 请求失败 (" + ms + "ms): " + e.getMessage());
                throw e;
            }

            long ms = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNs);
            Log.i(TAG, "← " + response.code() + " (" + ms + "ms) " + request.url());
            return response;
        }
    }

    public static Retrofit getClient() {
        if (retrofit == null) {
            HttpLoggingInterceptor bodyLog = new HttpLoggingInterceptor(msg ->
                    Log.d(TAG, "[HTTP] " + msg));
            bodyLog.setLevel(BuildConfig.DEBUG
                    ? HttpLoggingInterceptor.Level.BODY
                    : HttpLoggingInterceptor.Level.BASIC);

            OkHttpClient client = new OkHttpClient.Builder()
                    .connectTimeout(10, TimeUnit.SECONDS)
                    .readTimeout(30, TimeUnit.SECONDS)
                    .writeTimeout(15, TimeUnit.SECONDS)
                    .callTimeout(60, TimeUnit.SECONDS)
                    .retryOnConnectionFailure(true)
                    .addInterceptor(new TimingInterceptor())
                    .addInterceptor(bodyLog)
                    .build();

            retrofit = new Retrofit.Builder()
                    .baseUrl(BASE_URL)
                    .client(client)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();

            Log.i(TAG, "OkHttp 客户端已初始化, BASE_URL=" + BASE_URL);
        }
        return retrofit;
    }

    public static ApiService getApiService() {
        return getClient().create(ApiService.class);
    }
}

File: app/src/main/java/com/example/parentalcontrol/network/ApiClient.kt
package com.example.parentalcontrol.network

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.util.Log
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.net.ConnectException
import java.net.SocketTimeoutException
import java.net.UnknownHostException
import java.util.concurrent.TimeUnit

/**
 * API 客户端
 * 负责与自建服务器通信
 */
object ApiClient {

    private const val TAG = "ApiClient"

    // 服务器地址 - 华为云服务器
    // HTTP API 端口: 3000
    // WebSocket 端口: 3001
    const val BASE_URL = "http://139.9.176.191:3000/"
    const val WS_URL = "http://139.9.176.191:3001/"

    private var retrofit: Retrofit? = null
    private var apiService: ApiService? = null

    // 当前用户的 Token
    var authToken: String? = null

    fun getApiService(): ApiService {
        if (apiService == null) {
            val loggingInterceptor = HttpLoggingInterceptor { message ->
                Log.d(TAG, message)
            }.apply {
                level = HttpLoggingInterceptor.Level.BODY
            }

            val client = OkHttpClient.Builder()
                .addInterceptor(loggingInterceptor)
                .addInterceptor { chain ->
                    val requestBuilder = chain.request().newBuilder()
                    authToken?.let { token ->
                        requestBuilder.addHeader("Authorization", "Bearer $token")
                    }
                    chain.proceed(requestBuilder.build())
                }
                .connectTimeout(15, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .writeTimeout(30, TimeUnit.SECONDS)
                .retryOnConnectionFailure(true)
                .build()

            retrofit = Retrofit.Builder()
                .baseUrl(BASE_URL)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .build()

            apiService = retrofit!!.create(ApiService::class.java)
        }
        return apiService!!
    }

    /**
     * 检查网络连通性
     */
    fun isNetworkAvailable(context: Context): Boolean {
        val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val network = connectivityManager.activeNetwork ?: return false
        val capabilities = connectivityManager.getNetworkCapabilities(network) ?: return false
        return capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
    }

    /**
     * 解析网络异常为可读信息
     */
    fun parseNetworkError(error: Throwable): String {
        return when (error) {
            is SocketTimeoutException -> {
                Log.e(TAG, "连接超时: ${error.message}")
                "连接超时，请检查网络或服务器是否正常运行"
            }
            is ConnectException -> {
                Log.e(TAG, "连接被拒绝: ${error.message}")
                "无法连接到服务器（${error.message}），请确认服务器地址和端口是否正确"
            }
            is UnknownHostException -> {
                Log.e(TAG, "DNS解析失败: ${error.message}")
                "DNS解析失败，请检查网络连接或服务器地址是否正确"
            }
            is java.net.NoRouteToHostException -> {
                Log.e(TAG, "无法路由到主机: ${error.message}")
                "无法路由到服务器，请检查防火墙或网络配置"
            }
            is javax.net.ssl.SSLHandshakeException -> {
                Log.e(TAG, "SSL握手失败: ${error.message}")
                "SSL证书验证失败，请检查服务器证书配置"
            }
            else -> {
                Log.e(TAG, "未知网络错误: ${error.message}", error)
                "网络错误: ${error.message}"
            }
        }
    }

    /**
     * 更新服务器地址（需在 Application 初始化前调用）
     */
    fun updateBaseUrl(url: String) {
        retrofit = null
        apiService = null
    }
}
